Importing necessary libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

Reading the dataset with Pandas after mounting the dataset in google colab

In [2]:
#reading the dataset with a dataframe
df = pd.read_csv("final_preprocessed_dataset.csv")
df=df.drop('Unnamed: 0',1)

/tmp/ipykernel_56330/315725189.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df=df.drop('Unnamed: 0',1)


In [3]:
#eye balling the data
df.head(5)

Resource       APT Group  Entry Point  \
0  4d74c8da7274bb56bab01f47334a9551f57a6dcf  Equation Group      54299.0   
1  cc221465dac981f4934fef39d41ddb2e1d26299f          APT 21       7627.0   
2  a49718feddf874a62049809af51e858ad41897c4           APT 1      38026.0   
3  78c00614535b949770708069615764d0155cad76    Gorgon Group     396884.0   
4  44787ddf91b10291f8338590c5d99839040e1cd5          APT 21       6948.0   

   'ACTIVEDS.DLL'  'ADVAPI32.DLL'  'API-MS-WIN-CRT-HEAP-L1-1-0.DLL'  \
0               0               1                                 0   
1               0               0                                 0   
2               0               1                                 0   
3               0               1                                 0   
4               0               0                                 0   

   'API-MS-WIN-CRT-LOCALE-L1-1-0.DLL'  'API-MS-WIN-CRT-MATH-L1-1-0.DLL'  \
0                                   0                                 0   
1                                   0                                 0   
2                                   0                                 0   
3                                   0                                 0   
4                                   0                                 0   

   'API-MS-WIN-CRT-RUNTIME-L1-1-0.DLL'  'API-MS-WIN-CRT-STDIO-L1-1-0.DLL'  \
0                                    0                                  0   
1                                    0                                  0   
2                                    0                                  0   
3                                    0                                  0   
4                                    0                                  0   

   ...  'NEUTRAL SYS DEFAULT'  'NEUTRAL'  'PORTUGUESE BRAZILIAN'  'RUSSIAN'  \
0  ...                      0          1                       0          0   
1  ...                      0          0                       0          0   
2  ...                      0          0                       0          0   
3  ...                      0          1                       0          0   
4  ...                      0          0                       0          0   

   'SERBIAN DEFAULT'  'SORBIAN DEFAULT'  'SPANISH MODERN'  'SWEDISH'  \
0                  0                  0                 0          0   
1                  0                  0                 0          0   
2                  0                  0                 0          0   
3                  0                  0                 0          0   
4                  0                  0                 0          0   

   'TURKISH DEFAULT'  APTGroup  
0                  0         9  
1                  0         3  
2                  0         0  
3                  0        10  
4                  0         3  

[5 rows x 151 columns]

Dropping resource column and categorical APT group column

In [4]:
df = df.drop('Resource', 1)
df =df.drop('APT Group',1)

/tmp/ipykernel_56330/921168626.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df = df.drop('Resource', 1)
/tmp/ipykernel_56330/921168626.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df =df.drop('APT Group',1)


In [5]:
#eye balling information about our dataset
df.head(5)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2862 entries, 0 to 2861
Columns: 149 entries, Entry Point to APTGroup
dtypes: float64(1), int64(148)
memory usage: 3.3 MB


Sklearn Train Test Split Method : Splitting the data into training and testing data.

In [6]:
#the labelled column is Apt Group and the rest is features
label = np.array(df['APTGroup'])  # Target
features = df.drop('APTGroup',1)  # Input/independent varibale

/tmp/ipykernel_56330/3580316132.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  features = df.drop('APTGroup',1)


In [7]:
features.head()


Entry Point  'ACTIVEDS.DLL'  'ADVAPI32.DLL'  \
0      54299.0               0               1   
1       7627.0               0               0   
2      38026.0               0               1   
3     396884.0               0               1   
4       6948.0               0               0   

   'API-MS-WIN-CRT-HEAP-L1-1-0.DLL'  'API-MS-WIN-CRT-LOCALE-L1-1-0.DLL'  \
0                                 0                                   0   
1                                 0                                   0   
2                                 0                                   0   
3                                 0                                   0   
4                                 0                                   0   

   'API-MS-WIN-CRT-MATH-L1-1-0.DLL'  'API-MS-WIN-CRT-RUNTIME-L1-1-0.DLL'  \
0                                 0                                    0   
1                                 0                                    0   
2                                 0                                    0   
3                                 0                                    0   
4                                 0                                    0   

   'API-MS-WIN-CRT-STDIO-L1-1-0.DLL'  'ATL71.DLL'  'ATL80.DLL'  ...  \
0                                  0            0            0  ...   
1                                  0            0            0  ...   
2                                  0            0            0  ...   
3                                  0            0            0  ...   
4                                  0            0            0  ...   

   'NEUTRAL DEFAULT'  'NEUTRAL SYS DEFAULT'  'NEUTRAL'  \
0                  0                      0          1   
1                  0                      0          0   
2                  1                      0          0   
3                  0                      0          1   
4                  0                      0          0   

   'PORTUGUESE BRAZILIAN'  'RUSSIAN'  'SERBIAN DEFAULT'  'SORBIAN DEFAULT'  \
0                       0          0                  0                  0   
1                       0          0                  0                  0   
2                       0          0                  0                  0   
3                       0          0                  0                  0   
4                       0          0                  0                  0   

   'SPANISH MODERN'  'SWEDISH'  'TURKISH DEFAULT'  
0                 0          0                  0  
1                 0          0                  0  
2                 0          0                  0  
3                 0          0                  0  
4                 0          0                  0  

[5 rows x 148 columns]

In [8]:
label

array([ 9,  3,  0, ..., 10,  5, 10])

In [9]:
#choosing a 70-30 split to test out the performance
from sklearn.model_selection import train_test_split
seed =50
X_train, X_test, y_train, y_test = train_test_split(features,label,test_size=0.30, random_state = seed)

In [10]:
X_train.head(5)

Entry Point  'ACTIVEDS.DLL'  'ADVAPI32.DLL'  \
800       83741.0               0               1   
2684    1079966.0               0               0   
64        24146.0               0               1   
1010     180864.0               0               0   
2360     430686.0               0               0   

      'API-MS-WIN-CRT-HEAP-L1-1-0.DLL'  'API-MS-WIN-CRT-LOCALE-L1-1-0.DLL'  \
800                                  0                                   0   
2684                                 0                                   0   
64                                   0                                   0   
1010                                 0                                   0   
2360                                 0                                   0   

      'API-MS-WIN-CRT-MATH-L1-1-0.DLL'  'API-MS-WIN-CRT-RUNTIME-L1-1-0.DLL'  \
800                                  0                                    0   
2684                                 0                                    0   
64                                   0                                    0   
1010                                 0                                    0   
2360                                 0                                    0   

      'API-MS-WIN-CRT-STDIO-L1-1-0.DLL'  'ATL71.DLL'  'ATL80.DLL'  ...  \
800                                   0            0            0  ...   
2684                                  0            0            0  ...   
64                                    0            0            0  ...   
1010                                  0            0            0  ...   
2360                                  0            0            0  ...   

      'NEUTRAL DEFAULT'  'NEUTRAL SYS DEFAULT'  'NEUTRAL'  \
800                   0                      0          0   
2684                  0                      0          1   
64                    0                      0          0   
1010                  0                      0          0   
2360                  0                      0          1   

      'PORTUGUESE BRAZILIAN'  'RUSSIAN'  'SERBIAN DEFAULT'  'SORBIAN DEFAULT'  \
800                        0          0                  0                  0   
2684                       0          0                  0                  0   
64                         0          0                  0                  0   
1010                       0          0                  0                  0   
2360                       0          0                  0                  0   

      'SPANISH MODERN'  'SWEDISH'  'TURKISH DEFAULT'  
800                  0          0                  0  
2684                 0          0                  0  
64                   0          0                  0  
1010                 0          0                  0  
2360                 0          0                  0  

[5 rows x 148 columns]

In [11]:
X_test.head(5)

Entry Point  'ACTIVEDS.DLL'  'ADVAPI32.DLL'  \
1688      23536.0               0               1   
2455      10136.0               0               0   
2716     470572.0               0               1   
2296      10710.0               0               0   
578      549060.0               0               1   

      'API-MS-WIN-CRT-HEAP-L1-1-0.DLL'  'API-MS-WIN-CRT-LOCALE-L1-1-0.DLL'  \
1688                                 0                                   0   
2455                                 0                                   0   
2716                                 0                                   0   
2296                                 0                                   0   
578                                  0                                   0   

      'API-MS-WIN-CRT-MATH-L1-1-0.DLL'  'API-MS-WIN-CRT-RUNTIME-L1-1-0.DLL'  \
1688                                 0                                    0   
2455                                 0                                    0   
2716                                 0                                    0   
2296                                 0                                    0   
578                                  0                                    0   

      'API-MS-WIN-CRT-STDIO-L1-1-0.DLL'  'ATL71.DLL'  'ATL80.DLL'  ...  \
1688                                  0            0            0  ...   
2455                                  0            0            0  ...   
2716                                  0            0            0  ...   
2296                                  0            0            0  ...   
578                                   0            0            0  ...   

      'NEUTRAL DEFAULT'  'NEUTRAL SYS DEFAULT'  'NEUTRAL'  \
1688                  0                      0          0   
2455                  0                      0          0   
2716                  0                      0          1   
2296                  0                      0          0   
578                   0                      0          1   

      'PORTUGUESE BRAZILIAN'  'RUSSIAN'  'SERBIAN DEFAULT'  'SORBIAN DEFAULT'  \
1688                       0          0                  0                  0   
2455                       0          0                  0                  0   
2716                       0          0                  0                  0   
2296                       0          0                  0                  0   
578                        0          0                  0                  0   

      'SPANISH MODERN'  'SWEDISH'  'TURKISH DEFAULT'  
1688                 0          0                  0  
2455                 0          0                  0  
2716                 0          0                  0  
2296                 0          0                  0  
578                  0          0                  0  

[5 rows x 148 columns]

Building the Random Forest Classifier

In [12]:
rf_classifier = RandomForestClassifier(
                      min_samples_leaf=50,
                      n_estimators=150,
                      bootstrap=True,
                      oob_score=True,
                      n_jobs=-1,
                      random_state=seed,
                      max_features='auto')

In [13]:
rf_classifier.fit(X_train,y_train)  # Training Process

RandomForestClassifier(min_samples_leaf=50, n_estimators=150, n_jobs=-1,
                       oob_score=True, random_state=50)

In [14]:
y_pred = rf_classifier.predict(X_test)  # Prediction

In [15]:
print(y_pred)

[ 0 11 10 11 10  9 10  9  0  7  9 10  0 11 11  0  9  9  9  0  9  9 10  7
  7 11 11 10 10  0  0 11 11  9  0  8 10  0 11  7 11 10  0  0 11  0 11  0
  9 11 11  0  5  9 11  0 11 10 11 11  7  1  0  0 10  0 11 11 11  7 10  8
  0  9 11  9 10 11 11 10 10  0 11 11 11  0  9  0 11  8  7  9 11  0  9  7
 11 10 11  9  9  5 10  8 11 11 11  7 11  8 11 11  8 11  9  9  8  9  7 10
  7  1 11  0 10 11  7  5  0  0  0  9 11 10 11  0  7 10 11  0  0  0  0  7
 11 11 11  0  9 11  0 11 10 11  8 10  0  0 11 11  0  9 10 11 10  9 11  9
 10 11 11 11  7 11  9 11  9 11 11  0 10 11  0 10 11  0  9  0 11 10 11  0
  9 11 10  1  8 11  0 10  9  9 10 11  9  0 10 10  9  0 10  0 11 10 11 10
  0 10 10  9 11 10  9  0  5  0  0  0 11  0 10 11  7  9  9  0 11 11  1 11
 10  7  7 11  0 11 10 10  7 10  0  0 11  7 11 11 11 11  9  0 10 10  0 10
  0 11  8 11  0  8 10 11  0  0 11 10 10 11 11  7 11  7  7 10  0  9  0  9
  0 11 11  9  0  7 10 11  0  8  0  8  0  9  7 11  0  9  0 11  7  9 10 11
  9 10 11  0 11  9 11  0 11  9 10  0  0 11  0  0 11

Confusion Matrix

In [16]:
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[107   0   0   0   0   0   0   2   2   5   0  10]
 [  2  17   0   0   0   0   0   2   0   4   2  35]
 [  4   0   0   0   0   0   0   4   0   0   1   1]
 [  0   0   0   0   0   0   0   0   0   0   1  27]
 [ 18   0   0   0   0   0   0   1   0   1   2  24]
 [ 13   0   0   0   0  10   0   5   3   1  13  36]
 [ 29   0   0   0   0   0   0   1   0   1   0  19]
 [  9   0   0   0   0   0   0  43   0   2  17  15]
 [  1   0   0   0   0   0   0   0  39   0   1   0]
 [  0   0   0   0   0   0   0   0   0 113   0   0]
 [  2   0   0   0   0   0   0   0   0   2  90   0]
 [ 36   0   0   0   0   0   0   0   0   0   6  80]]


0.580908032596042

# Observation:

Accuracy acquired about 58% accuracy. Needs more hyperparameter tuning to increase the accuracy. Maybe contributing other factors along the way too can lead to better accuracy.

## Model Attempt#2 : Hyper Parameter tunning

In [17]:
"""
rf_classifier = RandomForestClassifier(
                      min_samples_leaf=50,
                      n_estimators=150,
                      bootstrap=True,
                      oob_score=True,
                      n_jobs=-1,
                      random_state=seed,
                      max_features='auto')
                      """
rf_classifier = RandomForestClassifier(
                      min_samples_leaf=3,
                      n_estimators=300,
                      bootstrap=True,
                      oob_score=True,
                      n_jobs=-1,
                      random_state=seed,
                      max_features='auto')

In [18]:
rf_classifier.fit(X_train, y_train)
y_pred = rf_classifier.predict(X_test)
print("accuracy ", accuracy_score(y_test, y_pred) )

accuracy  0.830034924330617


#Cross validation 

In [19]:
df.head(5)

Entry Point  'ACTIVEDS.DLL'  'ADVAPI32.DLL'  \
0      54299.0               0               1   
1       7627.0               0               0   
2      38026.0               0               1   
3     396884.0               0               1   
4       6948.0               0               0   

   'API-MS-WIN-CRT-HEAP-L1-1-0.DLL'  'API-MS-WIN-CRT-LOCALE-L1-1-0.DLL'  \
0                                 0                                   0   
1                                 0                                   0   
2                                 0                                   0   
3                                 0                                   0   
4                                 0                                   0   

   'API-MS-WIN-CRT-MATH-L1-1-0.DLL'  'API-MS-WIN-CRT-RUNTIME-L1-1-0.DLL'  \
0                                 0                                    0   
1                                 0                                    0   
2                                 0                                    0   
3                                 0                                    0   
4                                 0                                    0   

   'API-MS-WIN-CRT-STDIO-L1-1-0.DLL'  'ATL71.DLL'  'ATL80.DLL'  ...  \
0                                  0            0            0  ...   
1                                  0            0            0  ...   
2                                  0            0            0  ...   
3                                  0            0            0  ...   
4                                  0            0            0  ...   

   'NEUTRAL SYS DEFAULT'  'NEUTRAL'  'PORTUGUESE BRAZILIAN'  'RUSSIAN'  \
0                      0          1                       0          0   
1                      0          0                       0          0   
2                      0          0                       0          0   
3                      0          1                       0          0   
4                      0          0                       0          0   

   'SERBIAN DEFAULT'  'SORBIAN DEFAULT'  'SPANISH MODERN'  'SWEDISH'  \
0                  0                  0                 0          0   
1                  0                  0                 0          0   
2                  0                  0                 0          0   
3                  0                  0                 0          0   
4                  0                  0                 0          0   

   'TURKISH DEFAULT'  APTGroup  
0                  0         9  
1                  0         3  
2                  0         0  
3                  0        10  
4                  0         3  

[5 rows x 149 columns]

In [20]:
 from sklearn.model_selection import ShuffleSplit
 ss = ShuffleSplit(n_splits=20, test_size=0.30)
accuracy=0
for train_indices, test_indices in ss.split(df):
  #print("Train")
  #print(X_train)
  #print("Test")
  #print(X_test)
  
  train = df.iloc[train_indices]
  y_train= np.array(train['APTGroup'])
  X_train= train.drop('APTGroup',1)
  

  test = df.iloc[test_indices]
  y_test= np.array(train['APTGroup'])
  X_test= train.drop('APTGroup',1)
  
  rf_classifier.fit(X_train, y_train)
  y_pred = rf_classifier.predict(X_test)
  print("accuracy ", accuracy_score(y_test, y_pred) )
  accuracy+=accuracy_score(y_test, y_pred)

print(accuracy/20)


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8637044433349975


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8652021967049426


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8706939590614079


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8746879680479281


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.873689465801298


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8671992011982027


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.854717923115327


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8612081877184223


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8652021967049426


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8676984523215178


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8627059410883674


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8647029455816275


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8686969545681478


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8617074388417374


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8622066899650525


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8657014478282576


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8706939590614079


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8567149276085871


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.873689465801298


/tmp/ipykernel_56330/705216485.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train= train.drop('APTGroup',1)
/tmp/ipykernel_56330/705216485.py:17: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_test= train.drop('APTGroup',1)


accuracy  0.8761857214178732
0.866350474288567


# Observation 1:

Accuracy acquired : 86.62 % for n_splits=20 \\

# Another method for cross validation: cross_val_score


In [21]:
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier
print(cross_val_score(rf_classifier, features, label, cv=20, scoring ='accuracy').mean())

0.8420600233100233


# Observation 2:

Accuracy acquired : 84.2 % for n_splits=20 \\

## Important Feature Calculation

In [22]:
rf_classifier.fit(X_train,y_train)

RandomForestClassifier(min_samples_leaf=3, n_estimators=300, n_jobs=-1,
                       oob_score=True, random_state=50)

In [23]:
df.head(3)

Entry Point  'ACTIVEDS.DLL'  'ADVAPI32.DLL'  \
0      54299.0               0               1   
1       7627.0               0               0   
2      38026.0               0               1   

   'API-MS-WIN-CRT-HEAP-L1-1-0.DLL'  'API-MS-WIN-CRT-LOCALE-L1-1-0.DLL'  \
0                                 0                                   0   
1                                 0                                   0   
2                                 0                                   0   

   'API-MS-WIN-CRT-MATH-L1-1-0.DLL'  'API-MS-WIN-CRT-RUNTIME-L1-1-0.DLL'  \
0                                 0                                    0   
1                                 0                                    0   
2                                 0                                    0   

   'API-MS-WIN-CRT-STDIO-L1-1-0.DLL'  'ATL71.DLL'  'ATL80.DLL'  ...  \
0                                  0            0            0  ...   
1                                  0            0            0  ...   
2                                  0            0            0  ...   

   'NEUTRAL SYS DEFAULT'  'NEUTRAL'  'PORTUGUESE BRAZILIAN'  'RUSSIAN'  \
0                      0          1                       0          0   
1                      0          0                       0          0   
2                      0          0                       0          0   

   'SERBIAN DEFAULT'  'SORBIAN DEFAULT'  'SPANISH MODERN'  'SWEDISH'  \
0                  0                  0                 0          0   
1                  0                  0                 0          0   
2                  0                  0                 0          0   

   'TURKISH DEFAULT'  APTGroup  
0                  0         9  
1                  0         3  
2                  0         0  

[3 rows x 149 columns]

In [24]:
pd.set_option("display.max.rows", None)

In [25]:
# Important Features
feature_importance = pd.DataFrame(rf_classifier.feature_importances_,
                                   index = X_train.columns,
                                    columns=['importance']).sort_values('importance',ascending=False)
feature_importance
# summarize feature importance
#for i,v in enumerate(importance):
 #   print('Feature: %0d, Score: %.5f' % (i,v))

importance
Entry Point                            0.113523
'NEUTRAL'                              0.100925
'MSVCRT.DLL'                           0.071646
'WS2_32.DLL'                           0.060210
'WININET.DLL'                          0.059289
'USER32.DLL'                           0.041084
'ENGLISH US'                           0.034525
'CHINESE SIMPLIFIED'                   0.034205
'OLE32.DLL'                            0.033786
'CRYPT32.DLL'                          0.033353
'SHLWAPI.DLL'                          0.030107
'SHELL32.DLL'                          0.028224
'ADVAPI32.DLL'                         0.025700
'KERNEL32.DLL'                         0.024428
'COMCTL32.DLL'                         0.022368
'MSCOREE.DLL'                          0.020973
'VERSION.DLL'                          0.020969
'OLEAUT32.DLL'                         0.020491
'GDI32.DLL'                            0.020474
'MPR.DLL'                              0.019744
'WSOCK32.DLL'                          0.017153
'URLMON.DLL'                           0.016854
'IPHLPAPI.DLL'                         0.015249
'MFC42.DLL'                            0.011794
'KOREAN'                               0.011315
'COMDLG32.DLL'                         0.010083
'SECUR32.DLL'                          0.008599
'PSAPI.DLL'                            0.008123
'NETAPI32.DLL'                         0.007672
'MSVBVM60.DLL'                         0.007562
'MSVCP60.DLL'                          0.006378
'NEUTRAL DEFAULT'                      0.004552
'IMAGEHLP.DLL'                         0.004212
'ENGLISH UK'                           0.003665
'NETMGR.DLL'                           0.003554
'USERENV.DLL'                          0.003541
'WTSAPI32.DLL'                         0.003537
'WINMM.DLL'                            0.003404
'LZ32.DLL'                             0.003242
'WINSPOOL.DRV'                         0.003178
'DNSAPI.DLL'                           0.003165
'GDIPLUS.DLL'                          0.003146
'NTOSKRNL.EXE'                         0.003016
'RPCRT4.DLL'                           0.002877
'RUSSIAN'                              0.002691
'WINHTTP.DLL'                          0.001701
'JAPANESE DEFAULT'                     0.001473
'GERMAN'                               0.001271
'HAL.DLL'                              0.001102
'GLU32.DLL'                            0.000964
'OLEDLG.DLL'                           0.000940
'OPENGL32.DLL'                         0.000918
'OLEACC.DLL'                           0.000910
'TURKISH DEFAULT'                      0.000749
'IMM32.DLL'                            0.000723
'NTDLL.DLL'                            0.000691
'MSVCR80.DLL'                          0.000590
'KSECDD.SYS'                           0.000578
'FRENCH'                               0.000557
'AVICAP32.DLL'                         0.000436
'MSIMG32.DLL'                          0.000418
'SETUPAPI.DLL'                         0.000315
'MSVCR90.DLL'                          0.000199
'CHINESE TRADITIONAL'                  0.000163
'MFC42U.DLL'                           0.000153
'UXTHEME.DLL'                          0.000131
'FLTMGR.SYS'                           0.000125
'SPANISH MODERN'                       0.000104
'DSOUND.DLL'                           0.000078
'D3D9.DLL'                             0.000073
'DDRAW.DLL'                            0.000029
'JAPANESE NEUTRAL'                     0.000019
'OLEPRO32.DLL'                         0.000007
'ALBANIAN DEFAULT'                     0.000000
'DIVEHI NEUTRAL'                       0.000000
'WLDAP32.DLL'                          0.000000
'CRT.DLL'                              0.000000
'COMSVCS.DLL'                          0.000000
'ENGLISH CAN'                          0.000000
'ENGLISH AUS'                          0.000000
'CZECH DEFAULT'                        0.000000
'AZERI LATIN'                          0.000000
'CABINET.DLL'                          0.000000
'AVIF

Exporting the results into an excel file for better view

In [26]:
excel_data = feature_importance.copy()

In [27]:
excel_data.to_csv('important_features.csv', index=True)

In [28]:
%ls

final_preprocessed_dataset.csv  RF_Important_Features.ipynb
important_features.csv


In [29]:
#checking with shell command if there was a proper write in the new file created
! cat important_features.csv

,importance
Entry Point,0.1135227691584883
'NEUTRAL',0.10092466538612968
'MSVCRT.DLL',0.07164555543678403
'WS2_32.DLL',0.06021025096471999
'WININET.DLL',0.05928930765063745
'USER32.DLL',0.04108418304807252
'ENGLISH US',0.03452493129886506
'CHINESE SIMPLIFIED',0.034204879639839474
'OLE32.DLL',0.03378636990379905
'CRYPT32.DLL',0.0333534655140927
'SHLWAPI.DLL',0.03010702426593647
'SHELL32.DLL',0.02822373748924809
'ADVAPI32.DLL',0.025700081833147783
'KERNEL32.DLL',0.02442760731002417
'COMCTL32.DLL',0.022367992911429158
'MSCOREE.DLL',0.020972607134999037
'VERSION.DLL',0.020968523811916003
'OLEAUT32.DLL',0.02049054948634341
'GDI32.DLL',0.020473579698571456
'MPR.DLL',0.019744029319572195
'WSOCK32.DLL',0.01715251506450291
'URLMON.DLL',0.01685379289637437
'IPHLPAPI.DLL',0.015248501121645488
'MFC42.DLL',0.011793999790273013
'KOREAN',0.011314680757747152
'COMDLG32.DLL',0.01008277667709712
'SECUR32.DLL',0.008598728018898135
'PSAPI.DLL',0.008123330907142266
'NETAPI32.DLL',0.007671772120518975
'MSVB